Extension/src/LanguageServer/settings.ts

```diff
@@ -414,337 +424,253 @@ export class CppSettings extends Settings {
+        if (setting.enum !== undefined) {
+            if (this.isValidEnum(setting.enum, value)) {
+                return value;
+            }
+            return setting.default;
+        }
+        if (isString(value)) {
+            return value;
+        } else {
+            return undefined;
+        }
     }
 
-    public get vcFormatSpaceRemoveBeforeSemicolon(): boolean {
-        return super.Section.get<boolean>("vcFormat.space.removeBeforeSemicolon") === true;
+    // Returns the value of a setting as a boolean with proper type validation and checks for valid enum values while returning an undefined value if necessary.
+    private getAsBooleanOrUndefined(settingName: string): boolean | undefined {
+        const value: any = super.Section.get<any>(settingName);
+        if (isBoolean(value)) {
+            return value;
+        } else {
+            return undefined;
+        }
     }
 
-    public get vcFormatSpaceInsertAfterSemicolon(): boolean {
-        return super.Section.get<boolean>("vcFormat.space.insertAfterSemicolon") === true;
+    // Returns the value of a setting as a boolean with proper type validation.
+    private getAsBoolean(settingName: string): boolean;
+    private getAsBoolean(settingName: string, allowNull: boolean): boolean | null;
+    private getAsBoolean(settingName: string, allowNull: boolean = false): boolean | null {
+        const value: any = super.Section.get<any>(settingName);
+        if (allowNull && value === null) {
+            return null;
+        }
+        if (isBoolean(value)) {
+            return value;
+        }
+        const setting = getRawSetting("C_Cpp." + settingName);
+        return setting.default;
     }
 
-    public get vcFormatSpaceRemoveAroundUnaryOperator(): boolean {
-        return super.Section.get<boolean>("vcFormat.space.removeAroundUnaryOperator") === true;
-    }
+    // Returns the value of a setting as a string with proper type validation and checks for valid enum values.
+    private getAsString(settingName: string): string;
+    private getAsString(settingName: string, allowNull: boolean): string | null;
+    private getAsString(settingName: string, allowNull: boolean = false): string | null {
+        const value: any = super.Section.get<any>(settingName);
+        if (allowNull && value === null) {
+            return null;
+        }
+        const setting = getRawSetting("C_Cpp." + settingName);
 
-    public get vcFormatSpaceAroundBinaryOperator(): string {
-        // These strings have default values in package.json, so should never be undefined.
-        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        return super.Section.get<string>("vcFormat.space.aroundBinaryOperator")!;
+        if (setting.enum !== undefined) {
+            if (settingName === "loggingLevel" && isNumber(Number(value)) && Number(value) >= 0) {
+                return value;
+            }
+            if (this.isValidEnum(setting.enum, value)) {
```

Comment:
I would rather we reuse the `getNumericLoggingLevel` helper function here and test that the return value is >= 0

Reply:
Hmm, though that function expects a string and we haven't validated that at this point.  Maybe this is fine how it is.

Reply:
No, `value` is still `any` here so you can't pass that to a function that expects a `string`. That's what I was trying to say.

Reply:
And you must also check that the settingName is "loggingLevel".  That check should not have been removed.